gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,393 @@
+    } else {
+      dbNames = ImmutableList.of(databaseName(namespace));
+    }
+
+    ImmutableList.Builder<TableIdentifier> result = ImmutableList.builder();
+    dbNames.stream()
+        .map(
+            dbName ->
+                Streams.stream(bigLakeClient.listTables(dbName))
+                    .map(BigLakeCatalog::tableIdentifier)
+                    .collect(ImmutableList.toImmutableList()))
+        .forEach(result::addAll);
+    return result.build();
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    TableOperations ops = null;
+    TableMetadata lastMetadata = null;
+    if (purge) {
+      ops = newTableOps(identifier);
+      // TODO: to catch NotFoundException as in https://github.com/apache/iceberg/pull/5510.
+      lastMetadata = ops.current();
+    }
+
+    try {
+      bigLakeClient.deleteTable(tableName(databaseId(identifier.namespace()), identifier.name()));
+    } catch (NoSuchTableException e) {
+      return false;
+    }
+
+    if (ops != null && lastMetadata != null) {
+      CatalogUtil.dropTableData(ops.io(), lastMetadata);
+    }
+
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    String fromDbId = databaseId(from.namespace());
+    String toDbId = databaseId(to.namespace());
+
+    Preconditions.checkArgument(
+        fromDbId.equals(toDbId),
+        "Cannot rename table %s to %s: database must match",
+        from.toString(),
+        to.toString());
+    bigLakeClient.renameTable(tableName(fromDbId, from.name()), tableName(toDbId, to.name()));
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    if (namespace.isEmpty()) {
+      // Used by `CREATE NAMESPACE <catalog>`. Create a BLMS catalog linked with Iceberg catalog.
+      bigLakeClient.createCatalog(catalogName, Catalog.getDefaultInstance());
+      LOG.info("Created BigLake catalog: {}", catalogName.toString());
+    } else if (namespace.levels().length == 1) {
+      // Create a database.
+      String dbId = databaseId(namespace);
+      Database.Builder builder = Database.newBuilder().setType(Database.Type.HIVE);
+      builder
+          .getHiveOptionsBuilder()
+          .putAllParameters(metadata)
+          .setLocationUri(databaseLocation(dbId));
+
+      bigLakeClient.createDatabase(
+          DatabaseName.of(bigLakeProjectId, bigLakeRegion, catalogId, dbId), builder.build());
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Invalid namespace (too long): %s", namespace));
+    }
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    if (!namespace.isEmpty()) {
+      // BLMS does not support namespaces under database or tables, returns empty.
+      // It is called when dropping a namespace to make sure it's empty, returns empty to unblock
+      // deletion.
+      return ImmutableList.of();
```

Comment:
Shouldn't this check whether the namespace exists? If not then it should throw `NoSuchNamespaceException`.

Reply:
It should return empty list for database namespace, and throw NoSuchNamespaceException if the namespace is invalid (neither a catalog nor database). Updated it.

Reply:
I think that this needs to throw `NoSuchNamespaceException` if the database doesn't exist.